From 17ca42fb5fb624f03e278c56b72929dc202903bd Mon Sep 17 00:00:00 2001
From: Yu-Ming Chang <yumin686@andestech.com>
Date: Tue, 5 Sep 2023 10:25:36 +0800
Subject: [PATCH 057/241] target/riscv: The CSR mmsc_cfg2 now only exists in
 RV32.               Add RV32 write mask for CSR mecc_code.               Fix
 a veccfg() typo.

Signed-off-by: Yu-Ming Chang <yumin686@andestech.com>
---
 target/riscv/cpu.c       |  5 +++++
 target/riscv/csr_andes.c | 30 ++++++++++++++++++++++++++----
 target/riscv/csr_andes.h |  3 ++-
 3 files changed, 33 insertions(+), 5 deletions(-)

diff --git a/target/riscv/cpu.c b/target/riscv/cpu.c
index e26fdff441..62de710f79 100644
--- a/target/riscv/cpu.c
+++ b/target/riscv/cpu.c
@@ -904,6 +904,11 @@ static void rv64_andes_nx45v_cpu_init(Object *obj)
     set_satp_mode_max_supported(RISCV_CPU(obj), VM_1_10_MBARE);
 #endif
 
+    /* RVV extension */
+    cfg->vext_spec = g_strdup("v1.0");
+    cfg->vlen = 512;
+    cfg->elen = 64;
+
     /* Set CPU ID & Implementation ID */
     cfg->marchid = (uint64_t)0x1 << 63 | 0x8045;
     cfg->mimpid = 0x1;
diff --git a/target/riscv/csr_andes.c b/target/riscv/csr_andes.c
index c89c4a2327..95941f82c6 100644
--- a/target/riscv/csr_andes.c
+++ b/target/riscv/csr_andes.c
@@ -19,6 +19,15 @@ static RISCVException any(CPURISCVState *env,
     return RISCV_EXCP_NONE;
 }
 
+static RISCVException any32(CPURISCVState *env, int csrno)
+{
+    if (riscv_cpu_mxl(env) != MXL_RV32) {
+        return RISCV_EXCP_ILLEGAL_INST;
+    }
+
+    return any(env, csrno);
+}
+
 static RISCVException smode(CPURISCVState *env, int csrno)
 {
     if (riscv_has_ext(env, RVS)) {
@@ -153,7 +162,7 @@ static RISCVException veccfg(CPURISCVState *env, int csrno)
         }
         else {
             veccfgbit = get_field(csr->csrno[CSR_MMSC_CFG],
-                                  MASK_MMSC_CFG_RVARCH);
+                                  MASK_MMSC_CFG_VECCFG);
         }
         if (veccfgbit) {
             return RISCV_EXCP_NONE;
@@ -275,8 +284,21 @@ static RISCVException write_csr(CPURISCVState *env,
 static RISCVException write_mecc_code(CPURISCVState *env, int csrno,
                                       target_ulong val)
 {
-    env->andes_csr.csrno[CSR_MECC_CODE] = val & WRITE_MASK_CSR_MECC_CODE;
-    return RISCV_EXCP_NONE;
+    switch(riscv_cpu_mxl(env))
+    {
+        case MXL_RV32:
+            env->andes_csr.csrno[CSR_MECC_CODE] =
+                val & WRITE_MASK_CSR_MECC_CODE_32;
+            return RISCV_EXCP_NONE;
+            break;
+        case MXL_RV64:
+            env->andes_csr.csrno[CSR_MECC_CODE] =
+                val & WRITE_MASK_CSR_MECC_CODE_64;
+            return RISCV_EXCP_NONE;
+            break;
+        default:
+            return RISCV_EXCP_ILLEGAL_INST;
+    }
 }
 
 static RISCVException write_ucode(CPURISCVState *env, int csrno,
@@ -628,7 +650,7 @@ riscv_csr_operations andes_csr_ops[CSR_TABLE_SIZE] = {
     [CSR_MICM_CFG]     = { "micm_cfg",          any,    read_csr },
     [CSR_MDCM_CFG]     = { "mdcm_cfg",          any,    read_csr },
     [CSR_MMSC_CFG]     = { "mmsc_cfg",          any,    read_csr },
-    [CSR_MMSC_CFG2]    = { "mmsc_cfg2",         any,    read_csr },
+    [CSR_MMSC_CFG2]    = { "mmsc_cfg2",         any32,  read_csr },
     [CSR_MVEC_CFG]     = { "mvec_cfg",          veccfg, read_csr },
     [CSR_MRVARCH_CFG]  = { "mrvarch_cfg",       rvarch, read_csr },
 
diff --git a/target/riscv/csr_andes.h b/target/riscv/csr_andes.h
index f280bf89a6..db3ff14607 100644
--- a/target/riscv/csr_andes.h
+++ b/target/riscv/csr_andes.h
@@ -147,7 +147,8 @@ typedef struct AndesVec {
 #define MASK_MRVARCH_CFG_SVNAPOT            (0x1 << 24)
 #define MASK_MRVARCH_CFG_ZIHINTPAUSE        (0x1 << 25)
 
-#define WRITE_MASK_CSR_MECC_CODE            0xFF
+#define WRITE_MASK_CSR_MECC_CODE_32         0x7F
+#define WRITE_MASK_CSR_MECC_CODE_64         0xFF
 #define WRITE_MASK_CSR_UITB_32              0xFFFFFFFC
 #define WRITE_MASK_CSR_UITB_64              0xFFFFFFFFFFFFFFFC
 #define WRITE_MASK_CSR_MPFT_CTL             0x1F0
-- 
2.34.1

